datasync/chaindatafetcher/chaindata_fetcher.go

```diff
@@ -187,6 +193,24 @@ func (f *ChainDataFetcher) Stop() error {
 func (f *ChainDataFetcher) sendRequests(startBlock, endBlock uint64, reqType cfTypes.RequestType, shouldUpdateCheckpoint bool, stopCh chan struct{}) {
 	logger.Info("sending requests is started", "startBlock", startBlock, "endBlock", endBlock)
 	for i := startBlock; i <= endBlock; i++ {
+		for { // spin lock if processing data size is larger than max
+			select {
+			case <-stopCh:
+				logger.Info("stopped making requests", "startBlock", startBlock, "endBlock", endBlock, "stoppedBlock", i)
+				return
+			default:
+			}
+
+			f.lock.RLock()
+			if f.onProcessingDataSize <= f.maxProcessingDataSize {
+				f.lock.RUnlock()
+				break
+			}
+			f.lock.RUnlock()
+
+			logger.Debug("throttling the requests, sleeping", "interval", DefaultThrottlingInterval, "processingDataSize", f.onProcessingDataSize, "maxDatasize", f.maxProcessingDataSize)
+			time.Sleep(DefaultThrottlingInterval)
+		}
```

Comment:
Please take a look this refactored snippet
```diff
logger.Info("sending requests is started", "startBlock", startBlock, "endBlock", endBlock)
for i := startBlock; i <= endBlock; i++ {
    for { // spin lock if processing data size is larger than max
        select {
        case <-stopCh:
            logger.Info("stopped making requests", "startBlock", startBlock, "endBlock", endBlock, "stoppedBlock", i)
            return
        default:
            f.lock.RLock()
            if f.onProcessingDataSize <= f.maxProcessingDataSize {
                f.lock.RUnlock()
                f.reqCh <- cfTypes.NewRequest(reqType, shouldUpdateCheckpoint, i)
                break
            }
            f.lock.RUnlock()

            logger.Debug("throttling the requests, sleeping", "interval", DefaultThrottlingInterval, "processingDataSize", f.onProcessingDataSize, "maxDatasize", f.maxProcessingDataSize)
            time.Sleep(DefaultThrottlingInterval)
        }
    }
}
logger.Info("sending requests is finished", "startBlock", startBlock, "endBlock", endBlock)
```

If you agree, you need to slightly modify the test code like below
```
go func() {
    <-fetcher.reqCh
}()

// stop fetching while waiting
assert.NoError(t, fetcher.stopRangeFetching())
```

Reply:
@hyunsooda I disagreed with this sinppet. What if `f.reqCh` channel is full, but `stopCh` is closed?

Reply:
The suggested change and the existing one seem almost logically the same.

The existing one:
```
for {
   check stop signal
   wait or break
}
check stop signal
send request
```

The suggested one:
```
for {
  check stop signal
  send and break if throttling is not triggered, wait otherwise
}
```

The only difference is the existence of the additional checking stop signal, which seems redundant. Could you explain which part of the suggested one may be dangerous?